project-koku/koku
# COST-440 ocp summary presto

## Jira Ticket

[COST-440](https://issues.redhat.com/browse/COST-440)

## Description

OCP summarization via Presto.

This PR has **mass quantities** of changes to incorporate the processing of the raw line-item data via Presto and store _only_ the daily summary data in PostgreSQL.

![](https://media.giphy.com/media/l0MYvFhyRMgigx88U/giphy.gif)

## Breakdown

### Environment

Localdev settings:

```bash
export ENABLE_PARQUET_PROCESSING=True
export S3_BUCKET_NAME=koku-bucket
export S3_ENDPOINT=http://kokuminio:9000
export S3_ACCESS_KEY=kokuminioaccess
export S3_SECRET=kokuminiosecret
```

### Line Item Processing

* Process the raw label information (`tag=val<sep>...`) into json
* Ensure null label columns are translated to an empty json doc (`{}`)

### Presto Processing Code

* A wrapper module was created to process a buffer of statements that can be templated and pre-processed. This includes functions for:
    * connect
    * get cursor
    * execute _handles parameter substirution_
    * execute script _handles splitting script into individual statements and can execute a preprocessor against the statements_
* A cheap-n-cheerful mogrify function was created to allow for parameters to be properly injected into the SQL to be executed on Presto
* The Pod and Storage usage processing have been combined into a single summarization function call
* The Pod and Storage label/tag processing have been combined into a single summarization function call

### Presto SQL

* Processing involves creating some "temp" tables. These are actual tables prefixed with double-underscores that will be cleaned up after processing
* Daily tables are not updated. Processing uses the raw data and writes the final daily summarized data to PostgreSQL.
* An interface for deletions in PostgreSQL initiated from Presto was created
* An extension to that interface to allow for bulk delete for a simulated update (delete old, insert new) was created
* Presto has some problems with our high numeric precision columns. A table interface was created to handle this as well

### PostgreSQL

* The clone schema function will now copy rules and comments
* A table called `presto_delete_wrapper_log` was created to log delete actions that are submitted from presto. An insert into this table will activate a trigger that will then execute a delete on PostgreSQL using the data regarding the target table and conditions in the inserted record.
* A table called `presto_pk_delete_wrapper_log` was created to help with bulk delete using primary key values. The intent is to store primary key values in this table via a call from the presto processing. Then a delete record will be inserted into the `presto_delete_wrapper_log` table with conditions that reference this pk table for bulk delete.
* Since Presto has a problem where it will not even recognize our high precision columns from the daily summary table, a view was created (`reporting_ocpusagelineitem_daily_summary_presto`) where these numeric columns are changed to text type. This allows Presto to see the full table definition.
* A rule was created on this table that will intercept an insert into this view and will instead perform the insert on the actual summary table with the text columns cast to their proper data types for PostgreSQL storage.
* Table, column comments were added to these interfaces

## Testing

1. Setup the environment
2. Rebuild your containers and startup koku
3. `make docker-presto-up`
4. Wait ~30 seconds for all startup tasks to complete
5. `make load-test-customer-data`
6. Wait on that processing to finish
7. `presto --host=localhost --port=8080 --schema=acct10001 --user=admin --catalog=hive`
    1. `show tables` -- will list all available tables
    2. You can issue SQL statements to probe the OCP tables and see data present.
8. Logon to PostgreSQL
9. Use sql to probe the reporting_ocp tables. The daily table will not be updated, but the daily_summary and label summary tables will have data.
10. The label summary table(s) will  have all available labels in them
11. The daily_summary will only have enabled labels in them
12. koku-worker logs for these changes from the ocp_report_db_accessor will have entries prefixed with **PRESTO OCP**
